src/core/lib/Bech32.mjs

```diff
@@ -0,0 +1,286 @@
+ * @param {string} pre
+ * @returns
+ */
+function polymodStep (pre) {
+    const b = pre >> 25;
+    return ((pre & 0x1FFFFFF) << 5) ^
+    (-((b >> 0) & 1) & 0x3b6a57b2) ^
+    (-((b >> 1) & 1) & 0x26508e6d) ^
+    (-((b >> 2) & 1) & 0x1ea119fa) ^
+    (-((b >> 3) & 1) & 0x3d4233dd) ^
+    (-((b >> 4) & 1) & 0x2a1462b3);
+}
+
+/**
+ * Checks the prefix of a string.
+ * @param {*} prefix
+ * @returns
+ */
+function prefixChk (prefix) {
+    let chk = 1;
+    for (let i = 0; i < prefix.length; ++i) {
+        const c = prefix.charCodeAt(i);
+        if (c < 33 || c > 126) return "KO";
+        chk = polymodStep(chk) ^ (c >> 5);
+    }
+    chk = polymodStep(chk);
+
+    for (let i = 0; i < prefix.length; ++i) {
+        const v = prefix.charCodeAt(i);
+        chk = polymodStep(chk) ^ (v & 0x1f);
+    }
+    return chk;
+}
+
+/**
+ * Bech32 Checksum
+ * We check the entire string to see if its segwit encoded.
+ * Lengths and other constants taken from BIP 0173: https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki
+ *
+ * @param {*} str
+ * @returns
+ */
+function checkbech32 (str) {
+    const LIMIT = 90;
+    if (str.length < 8) return "KO";
+    if (str.length > LIMIT) return "KO";
+
+
+    const split = str.lastIndexOf("1");
+    if (split === -1) return "KO";
+    if (split === 0) return "KO";
+
+    const prefix = str.slice(0, split);
+    const wordChars = str.slice(split + 1);
+    if (wordChars.length < 6) return "KO";
+
+    let chk = prefixChk(prefix);
+    if (typeof chk === "string") return "KO";
+
+    const words = [];
+    for (let i = 0; i < wordChars.length; ++i) {
+        const c = wordChars.charAt(i);
+        const v = ALPHABET_MAP[c];
+        if (v === undefined) return "KO";
+        chk = polymodStep(chk) ^ v;
+        if (i + 6 >= wordChars.length) continue;
+        words.push(v);
+    }
+    // Second number is decimal representation of 0x2bc830a3
+    // Useful as P2TR addresses are segwit encoded, with different final checksum.
+    // Taken from https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki
+    if (chk === 1 || chk === 734539939) {
+        return "OK";
+    } else {
+        return "KO";
+    }
+}
+
+// ################################################ END SEGWIT DECODING FUNCTIONS ###################################################
+
+// ################################################ BEGIN MAIN CHECKSUM FUNCTIONS ###################################################
+
+// Segwit Checksum
+/**
+ * Segwit Checksum
+ * @param {*} str
+ * @returns
+ */
+export function segwitChecksum(str) {
+    return (checkbech32(str) === "OK");
+}
+
+// ################################################ END MAIN CHECKSUM FUNCTIONS #####################################################
+
+
+// ################################################ BEGIN SEGWIT ENCODING FUNCTIONS #################################################
+
+// Taken from https://github.com/sipa/bech32/blob/master/ref/javascript/bech32.js
```

Comment:
Similar to above, this source includes an MIT license that wants to be copied over as well.